src/scanpy/preprocessing/_simple.py

```diff
@@ -612,6 +612,47 @@ def normalize_per_cell(  # noqa: PLR0917
     return X if copy else None
 
 
+@numba.njit(cache=True, parallel=True)
+def to_dense(shape, indptr, indices, data):
+    """\
+    Numba kernel for np.toarray() function
+    """
+    X = np.empty(shape, dtype=data.dtype)
+
+    for r in numba.prange(shape[0]):
+        X[r] = 0
+        for i in range(indptr[r], indptr[r + 1]):
+            X[r, indices[i]] = data[i]
+    return X
+
+
+def numpy_regress_out(
+    X: np.ndarray,
+    A: np.ndarray,
+) -> np.ndarray:
+    """\
+    Numba kernel for regress out unwanted sorces of variantion.
+    Finding coefficient using Linear regression (Linear Least Squares).
+    """
+
+    @numba.njit(cache=True, parallel=True)
+    def get_resid(
+        X: np.ndarray,
+        A: np.ndarray,
+        coeff,
+    ) -> np.ndarray:
+        for i in numba.prange(X.shape[0]):
+            X[i] -= A[i] @ coeff
+        return X
+
+    tmp0 = A.T @ A
+    tmp = np.linalg.inv(tmp0)
+    tmp0 = A.T @ X
+    coeff = tmp @ tmp0
```

Comment:
ditto, `tmp0` and `tmp` aren’t good names, and re-declaring them makes this more confusing.

E.g. this would be a start:

```suggestion
    todo_rename_me = np.linalg.inv(A.T @ A)
    coeff = todo_rename_me @ (A.T @ X)
```